File: Example2018/Example2018/ViewController.m
//
//  ViewController.m
//  Example2018
//
//  Created by Eric Larson on 8/21/18.
//  Copyright © 2018 Eric Larson. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()
@property (weak, nonatomic) IBOutlet UILabel *ericLabel;
@property (strong, nonatomic) NSNumber* number;
@end

@implementation ViewController
@synthesize number = _number;

-(NSNumber*)number{
    if(!_number){
        _number = @1;
    }
    return _number;
}

-(void)setNumber:(NSNumber *)number{
    _number = number;
    self.ericLabel.text = [NSString stringWithFormat:@"%.2f",[self.number floatValue]];
    //self.number
}

- (void)viewDidLoad {
    [super viewDidLoad];
}

- (IBAction)didPressButton:(UIButton *)sender {
    [self changeLabel:self.ericLabel withString:@"Eric"];
}

- (void)changeLabel:(UILabel* )label
         withString:(NSString*)string
{
    label.text = string;
}
- (IBAction)sliderChanged:(UISlider *)sender {
    self.number = @(sender.value);
}


@end

File: ICA4/ExampleRedBearChat/ViewController.swift
//
//  ViewController.swift
//  ExampleRedBearChat
//
//  Created by Eric Larson on 9/26/17.
//  Copyright © 2017 Eric Larson. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    // MARK: ====== VC Properties ======
    lazy var bleShield:BLE = (UIApplication.shared.delegate as! AppDelegate).bleShield
    var rssiTimer = Timer()
    @IBOutlet weak var spinner: UIActivityIndicatorView!
    @IBOutlet weak var labelText: UILabel!
    @IBOutlet weak var textBox: UITextField!
    @IBOutlet weak var rssiLabel: UILabel!
    @IBOutlet weak var connectLabel: UILabel!

    override func viewDidLoad() {
        super.viewDidLoad()

        // BLE Connect Notification
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(self.onBLEDidConnectNotification),
                                               name: NSNotification.Name(rawValue: kBleConnectNotification),
                                               object: nil)

        // BLE Disconnect Notification
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(self.onBLEDidDisconnectNotification),
                                               name: NSNotification.Name(rawValue: kBleDisconnectNotification),
                                               object: nil)

        // BLE Recieve Data Notification
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(self.onBLEDidRecieveDataNotification),
                                               name: NSNotification.Name(rawValue: kBleReceivedDataNotification),
                                               object: nil)
    }

    func readRSSITimer(timer:Timer){
        bleShield.readRSSI { (number, error) in
            // when RSSI read is complete, display it
            self.rssiLabel.text = String(format: "%.1f",(number?.floatValue)!)
            self.spinner.stopAnimating()
        }
    }

    @objc func onBLEDidConnectNotification(notification:Notification){
        print("Notification arrived that BLE Connected")

        let deviceName = notification.userInfo?["name"] as? String
        self.connectLabel.text = deviceName
        self.spinner.startAnimating()
        rssiTimer = Timer.scheduledTimer(withTimeInterval: 1.0,
                                         repeats: true,
                                         block: self.readRSSITimer)
    }

    // NEW  DISCONNECT FUNCTION
    @objc func onBLEDidDisconnectNotification(notification:Notification){
        print("Notification arrived that BLE Disconnected a Peripheral")

        self.connectLabel.text = "Disconnected"
        rssiTimer.invalidate()
    }

    // NEW FUNCTION EXAMPLE: this was written for you to show how to change to a notification based model
    @objc func onBLEDidRecieveDataNotification(notification:Notification){
        let d = notification.userInfo?["data"] as! Data?
        let s = String(bytes: d!, encoding: String.Encoding.utf8)
        self.labelText.text = s
    }

    // MARK: CHANGE: this function only needs a name change, the BLE writing does not change
    @IBAction func sendDataButton(_ sender: UIButton) {

        let s = textBox.text!
        let d = s.data(using: String.Encoding.utf8)!
        bleShield.write(d)
        // if (self.textField.text.length > 16)
    }

}








